selfdrive/modeld/modeld.py

```diff
@@ -0,0 +1,307 @@
+
+  state = PublishState()
+  params = Params()
+
+  # setup filter to track dropped frames
+  frame_dropped_filter = FirstOrderFilter(0., 10., 1. / MODEL_FREQ)
+  frame_id = 0
+  last_vipc_frame_id = 0
+  run_count = 0
+  # last = 0.0
+
+  model_transform_main = np.zeros((3, 3), dtype=np.float32)
+  model_transform_extra = np.zeros((3, 3), dtype=np.float32)
+  nav_enabled = False
+  live_calib_seen = False
+  driving_style = np.array([1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0], dtype=np.float32)
+  nav_features = np.zeros(NAV_FEATURE_LEN, dtype=np.float32)
+  nav_instructions = np.zeros(NAV_INSTRUCTION_LEN, dtype=np.float32)
+  buf_main, buf_extra = None, None
+  meta_main = FrameMeta()
+  meta_extra = FrameMeta()
+
+  while True:
+    # Keep receiving frames until we are at least 1 frame ahead of previous extra frame
+    while meta_main.timestamp_sof < meta_extra.timestamp_sof + 25000000:
+      buf_main = vipc_client_main.recv()
+      meta_main = FrameMeta(vipc_client_main)
+      if buf_main is None:
+        break
+
+    if buf_main is None:
+      logging.error("vipc_client_main no frame")
+      continue
+
+    if use_extra_client:
+      # Keep receiving extra frames until frame id matches main camera
+      while True:
+        buf_extra = vipc_client_extra.recv()
+        meta_extra = FrameMeta(vipc_client_extra)
+        if buf_extra is None or meta_main.timestamp_sof < meta_extra.timestamp_sof + 25000000:
+          break
+
+      if buf_extra is None:
+        logging.error("vipc_client_extra no frame")
+        continue
+
+      if abs(meta_main.timestamp_sof - meta_extra.timestamp_sof) > 10000000:
+        logging.error("frames out of sync! main: {} ({:.5f}), extra: {} ({:.5f})".format(
+          meta_main.frame_id, meta_main.timestamp_sof / 1e9,
+          meta_extra.frame_id, meta_extra.timestamp_sof / 1e9))
+
+    else:
+      # Use single camera
+      buf_extra = buf_main
+      meta_extra = meta_main
+
+    # TODO: path planner timeout?
+    sm.update(0)
+    desire = sm["lateralPlan"].desire.raw
+    is_rhd = sm["driverMonitoringState"].isRHD
+    frame_id = sm["roadCameraState"].frameId
+    if sm.updated["liveCalibration"]:
+      device_from_calib_euler = np.array(sm["liveCalibration"].rpyCalib, dtype=np.float32)
+      model_transform_main = update_calibration(device_from_calib_euler, main_wide_camera, False)
+      model_transform_extra = update_calibration(device_from_calib_euler, True, True)
+      live_calib_seen = True
+
+    traffic_convention = np.zeros(2)
+    traffic_convention[int(is_rhd)] = 1
+
+    vec_desire = np.zeros(DESIRE_LEN, dtype=np.float32)
+    if desire >= 0 and desire < DESIRE_LEN:
+      vec_desire[desire] = 1
+
+    # Enable/disable nav features
+    timestamp_llk = sm["navModel"].locationMonoTime
+    nav_valid = sm.valid["navModel"] # and (nanos_since_boot() - timestamp_llk < 1e9)
+    use_nav = nav_valid and params.get_bool("ExperimentalMode")
+    if not nav_enabled and use_nav:
```

Comment:
Why isn't this just:

if use_nav:
else:

Reply:
`nav_features` and `nav_instructions` only need to be zero'ed out on the falling edge of `use_nav`. Zero-ing them out on every frame wastes a bit of CPU.

Reply:
If it's a negligible amount of cpu it seems better to just set it every time, especially since you need that cpu anyway when nav is running.